core/src/OpenACC/Kokkos_OpenACC_Instance.cpp

```diff
@@ -0,0 +1,205 @@
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+// FIXME_OPENACC - macro for workaround implementation in UniqueToken
+// constructor. undef'ed at the end
+#define KOKKOS_IMPL_OPENACC_WORKAROUND
+
+#include <Kokkos_Core.hpp>
+
+#include <Kokkos_OpenACC.hpp>
+#include <OpenACC/Kokkos_OpenACC_Instance.hpp>
+#include <OpenACC/Kokkos_OpenACC_Exec.hpp>
+
+#include <vector>
+#include <ostream>
+#include <string>
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+uint32_t OpenACCInternal::impl_get_instance_id() const noexcept {
+  return m_instance_id;
+}
+
+void OpenACCInternal::fence(openacc_fence_is_static is_static) {
+  fence(
+      "Kokkos::Experimental::Impl::OpenACCInternal::fence: Unnammed Internal "
+      "Fence",
+      is_static);
+}
+void OpenACCInternal::fence(const std::string& name,
+                            openacc_fence_is_static is_static) {
+  if (is_static == openacc_fence_is_static::no) {
+    Kokkos::Tools::Experimental::Impl::profile_fence_event<
+        Kokkos::Experimental::OpenACC>(
+        name,
+        Kokkos::Tools::Experimental::Impl::DirectFenceIDHandle{
+            impl_get_instance_id()},
+        [&]() {
+          //[DEBUG] disabled due to synchronous behaviors of the current
+          // parallel construct implementations. acc_wait_all();
+        });
+  } else {
+    Kokkos::Tools::Experimental::Impl::profile_fence_event<
+        Kokkos::Experimental::OpenACC>(
+        name,
+        Kokkos::Tools::Experimental::SpecialSynchronizationCases::
+            GlobalDeviceSynchronization,
+        [&]() {  //[TODO]: correct device ID
+          acc_wait_all();
+        });
+  }
+}
+int OpenACCInternal::concurrency() { return 128000; }
+const char* OpenACCInternal::name() { return "OpenACC"; }
+void OpenACCInternal::print_configuration(std::ostream& os, const bool) {
+  // FIXME_OPENACC
+  os << "Using OpenACC\n";
+}
+
+void OpenACCInternal::impl_finalize() {
+  m_is_initialized = false;
+  Kokkos::Impl::OpenACCExec space;
+  if (space.m_lock_array != nullptr) space.clear_lock_array();
```

Comment:
Why have both `OpenACCInternal` and `OpenACCExec`?
Why not move this functionality into `OpenACCInternal`?

Reply:
OpenACCExec is merged into OpenACCInternal, and OpenACCExec is deleted.